### ISSUE TITLE
Is there any guide for implementing this?

### ISSUE THREAD

**Issue by paltaa** [pemcg/multi-level-approval#1]
Hey, ive been looking everywhere for the past week for a how-to develop this feature in automation, eventually found this repo but dont know how to implement this. 

Also a heads up on where to start to develop this feature would be really nice

**Comment by pemcg** [comment]
Hi

I'm not sure that I'd recommend this in its current state, it's pretty hacky. I put it together as a personal PoC to see if it was possible to get 3-way (expandable to n-way) approval working purely using automate code - no changes to Rails. It kinda works but the ugliness put me off developing it further.

The initial and final approvals are pretty straightforward, inserting the middle approval added to the ugliness quite a lot. I used a slightly modified version of one of Kevin Morey's CloudForms Essentials (https://github.com/ramrexx/CloudForms_Essentials) dialogs and correspondingly tweaked his build_vm_provision_request method. 

If you want to investigate further it uses 3 separate approval workflows: 

The service request (request approval workflow 1) launches an instance/method that runs an $evm.execute('create_automation_request'..) (request approval workflow 2) that runs Kevin's build_vm_provision_request (request approval workflow 3) to provision the actual VM.

If you just need 2-way approvals, stripping out the intermediate step would clean it up quite a bit, so you'd just have the service request approval and the provision request approval - fairly out-of-the-box stuff.

Much of the ugliness came from the edits to the various email methods to send the correct email text in the case of a 'normal' provision, or a multi-approval service. I'm sure that could be re-done in a much cleaner way, although the email classes have all now been consolidated in the current version of CloudForms.

To use this code you need to import the automate domain and service dialog. Create a new service of item type 'Generic' that uses the dialog. The dialog uses an invisible dialog element called 'approval_level' that's referenced all the way through the workflow (that's just one of the nasty hacks). Then you need to edit all of the email classes in the new domain to set the correct email addresses for the various approvers. I think the defaults are first_approver@my_org.com, second_approver@my_org.com and final_approver@my_org.com.

Then just order the service. The first approver should get an email. Approve that, and the second approver should get an email, but automation request should go into a pending state, and you have to approve it using the approve_request_via_api.rb script. Once that's approved the third approver should get an email, and once that is approved the VM should be created.

Hope this helps,
pemcg

**Comment by paltaa** [comment]
Peter,

Thank you for the fast and complete answer to my question, a 2 way approval is sufficient, so im going to try implementing this with your instructions.

**Comment by paltaa** [comment]
@pemcg Is there a possibility for you to help me implement this by slack or any other chat? im still not quite familiar with RoR and its been hard to understand every change need to be made in order to make this work, got an overall idea but still quite confused, im using a manageIQ fine version, thanks in advance,

JPQ

**Comment by suhasdombale123** [comment]
any progress on this?How to implement?

**Comment by paltaa** [comment]
Nope, left my last job so not my problem anymore haha

**Comment by suhasdombale123** [comment]
@pemcg is it possible to do second level approval???

**Comment by pemcg** [comment]
Yes you could leave out the intermediate request and you'd just have 2-way approval

**Comment by suhasdombale123** [comment]
Thanks but by leaving out the intermediate request do you mean just removing the intermediate request classes and methods?

**Comment by pemcg** [comment]
> The service request (request approval workflow 1) launches an instance/method that runs an $evm.execute('create_automation_request'..) (request approval workflow 2) that runs Kevin's build_vm_provision_request (request approval workflow 3) to provision the actual VM.

Each request includes an approval stage, so if you only wanted 2-way approval then you only need the initial service provision request (request approval workflow 1), and the final build_vm_provision_request (request approval workflow 2) to provision the actual VM. You'd have to re-factor the code a bit but it should be fairly straightforward. 

As mentioned though that this code was only done as a rough proof-of-concept and needs a lot of work to make production-ready.
